### DISCUSSION TITLE
How to add custom type conversion rule?

### DISCUSSION THREAD

**Discussion by evan0greenup** [PyO3/pyo3#3230] [Questions]
For my project, if I want to add custom conversion rules between Python type and Rust type or Vice versa. Is there any official support or recommend way to do that?

**Comment by davidhewitt** [comment]
Usually implementing the `FromPyObject` and `IntoPy<PyObject>` traits will give you custom conversions.

**Reply by jnoortheen** [reply]
@davidhewitt  Could you point to a complex example of implementing these traits. I am new to rust and currently wrapping the struct with `#[pyclass]` and duplicating the methods from the inner struct. 

https://github.com/jnoortheen/xonsh-parser/blob/3f2a297e9a7bac93d9b00f317d5ca63058074f8c/xonsh_tokenizer/src/lib.rs#L14

**Reply by davidhewitt** [reply]
I'd suggest looking at the files inside PyO3's own `src/conversions` directory.

But it's worth noting that `#[pyclass]` will automatically implement for you `FromPyObject for PyRef<T>` as well as `IntoPy<PyObject>`.

**Reply by jnoortheen** [reply]
Thanks @davidhewitt , I think the `src/conversions` and the `cargo expand` of the command would suffice to export the `TokInfo` struct as a `PyClass`.
